leader/clique/humblet.go

```diff
@@ -0,0 +1,235 @@
+package clique
+
+import (
+	"log"
+	"bytes"
+	"encoding/binary"
+	"encoding/json"
+	"github.com/krzysztof-turowski/distributed-framework/lib"
+)
+
+const (
+	msgCapture = iota
+	msgAccept
+	msgYes
+	msgNo
+	msgLeader
+)
+
+const none = -1;
+
+type stateHumblet struct {
+	Active    bool
+	Level     int
+	Owner     int
+	Queue     []captureMessage
+	Contender int
+	Leader    int
+}
+
+type captureMessage struct {
+	Sender int
+	Passed bool
+	Level  int
+	Id     int
+}
+
+func RunHumblet(nodes []lib.Node, runner lib.Runner) int {
+	for _, node := range nodes {
+		log.Println("Running node", node.GetIndex())
+		go runHumblet(node)
+	}
+
+	runner.Run()
+	runner.GetStats()
+
+	checkSingleLeaderElected(nodes)
+
+	leader := getState(nodes[0]).Leader
+	log.Println("Elected node", leader, "as a leader")
+	return leader
+}
+
+func runHumblet(node lib.Node) {
+	node.StartProcessing()
+
+	initializeHumblet(node)
+
+	for {
+		sender, message := node.ReceiveAnyMessage()
+		if processHumblet(node, sender, message) {
+			break
+		}
+	}
+
+	node.FinishProcessing(true)
+}
+
+func initializeHumblet(node lib.Node) {
+	setState(node, &stateHumblet{
+		Active: true,
+		Level: 0,
+		Owner: none,
+		Queue: make([]captureMessage, 0),
+		Contender: none,
+		Leader: none,
+	})
+
+	id := node.GetIndex()
+	log.Println("Node", id, "tries to capture first node")
+	node.SendMessage(0, encodeAll(byte(msgCapture), int32(0), int64(id)))
+}
+
+func processHumblet(node lib.Node, sender int, message []byte) bool {
+	id := node.GetIndex()
+	size := node.GetSize()
+	state := getState(node)
+
+	defer setState(node, &state)
+
+	captureNext := func() {
+		node.SendMessage(state.Level, encodeAll(byte(msgCapture), int32(state.Level), int64(id)))
+	}
+
+	passToOwner := func(message captureMessage) {
+		node.SendMessage(state.Owner, encodeAll(byte(msgCapture), int32(message.Level), int64(message.Id)))
+	}
+
+	respond := func(receiver int, label byte) {
+		node.SendMessage(receiver, encodeAll(label))
+	}
+
+	announceAsLeader := func() {
+		message := encodeAll(byte(msgLeader), int64(id))
+		for receiver := 0; receiver < size - 1; receiver++ {
+			node.SendMessage(receiver, message)
+		}
+	}
+
+	buffer := bytes.NewBuffer(message)
+
+	var label byte
+	decode(buffer, &label)
+
+	switch label {
+	case msgCapture:
+		passed := sender < state.Level
+		var level int32
+		var id int64
+		decode(buffer, &level, &id)
+		state.Queue = append(state.Queue, captureMessage{Sender: sender, Passed: passed, Level: int(level), Id: int(id)})
```

Comment:
Because `encoding/binary` package only supports fixed-size types. I kept manual binary encoding for messages, because it is more compact than JSON, and perhaps we could also want to measure the total number of bytes sent.